ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/ApplyMigrationCommand.java

```diff
@@ -311,22 +313,80 @@ private void executeCommands(
       final Clock clock,
       final String previous
   ) {
+    final List<String> commands = CommandParser.splitSql(migrationFileContent);
+
+    executeCommands(
+        commands, ksqlClient, config, executionStart, migration, clock, previous, true);
+    executeCommands(
+        commands, ksqlClient, config, executionStart, migration, clock, previous, false);
+  }
+
+  /**
+   * If validateOnly is set to true, then this parses each of the commands but only executes
+   * DEFINE/UNDEFINE commands (variables are needed for parsing INSERT INTO... VALUES, SET/UNSET
+   * and DEFINE commands). If validateOnly is set to fales, then each command will execute after
+   * parsing.
+   */
+  private void executeCommands(
+      final List<String> commands,
+      final Client ksqlClient,
+      final MigrationConfig config,
+      final String executionStart,
+      final MigrationFile migration,
+      final Clock clock,
+      final String previous,
+      final boolean validateOnly
+  ) {
+    cleanUpJavaClientVariables(ksqlClient);
     final Map<String, Object> properties = new HashMap<>();
-    final List<SqlCommand> commands = CommandParser.parse(migrationFileContent);
-    for (final SqlCommand command : commands) {
+    for (final String command : commands) {
       try {
-        executeCommand(command, ksqlClient, properties);
-      } catch (InterruptedException | ExecutionException e) {
+        final Map<String, String> variables = ksqlClient.getVariables().entrySet()
+            .stream().collect(Collectors.toMap(e -> e.getKey(), e -> e.getValue().toString()));
+        executeCommand(
+            CommandParser.transformToSqlCommand(command, variables),
+            ksqlClient,
+            properties,
+            validateOnly
+        );
+      } catch (InterruptedException | ExecutionException | MigrationException e) {
+        final String action = validateOnly ? "parse" : "execute";
         final String errorMsg = String.format(
-            "Failed to execute sql: %s. Error: %s", command.getCommand(), e.getMessage());
+            "Failed to %s sql: %s. Error: %s", action, command, e.getMessage());
         updateState(config, ksqlClient, MigrationState.ERROR,
             executionStart, migration, clock, previous, Optional.of(errorMsg));
         throw new MigrationException(errorMsg);
       }
     }
   }
 
+  private void cleanUpJavaClientVariables(final Client ksqlClient) {
+    ksqlClient.getVariables().forEach((k, v) -> ksqlClient.undefine(k));
+  }
+
   private void executeCommand(
+      final SqlCommand command,
+      final Client ksqlClient,
+      final Map<String, Object> properties,
+      final boolean defineUndefineOnly
+  ) throws ExecutionException, InterruptedException {
+    if (!defineUndefineOnly) {
```

Comment:
nit: can we move this into an `else` block for the if-else stack below to improve readability?
```
if (command instanceof SqlDefineVariableCommand) {
  ...
} else if (command instanceof SqlUndefineVariableCommand) {
  ...
} else {
  if (!defineUndefineOnly) {
    executeNonVariableCommands(command, ksqlClient, properties);
  }
}
```